parser/src/ast/module.rs

```diff
@@ -502,17 +502,18 @@ impl Module {
                     prev.span(),
                     segment.name.span(),
                 );
-                return Err(ModuleError::NameConflict(segment.name.span()));
+                return Err(SemanticAnalysisError::NameConflict(segment.name.span()));
             }
             for binding in segment.bindings.iter() {
-                if let Some(prev) = names.replace(NamespacedIdentifier::Binding(binding.name)) {
+                let binding_name = binding.name.unwrap();
```

Comment:
Yeah in this case it is impossible for the name to be `None`, since it is one of the trace segment bindings, which by definition always have names. I can add a comment here, but there are a number of places where this assumption is made. Maybe it would be worthwhile to discuss how to approach this sort of thing in the general case? I suspect I may be approaching these areas of the design/solution space with a set of criteria/tradeoffs that is out of sync with the rest of the team, which is why we are bumping in to this more frequently in the code I write. 

For example, my use of `unwrap` here (and in general) is based on these criteria:

1. An effort to reuse structures across contexts, where usage in those contexts is virtually indistinguishable. For example, using data structures across multiple phases of compilation, where later phases are allowed to assume certain fields to be set and/or properties to hold true. These assumptions are made explicit by using `unwrap`, `expect` or other assertions where needed. The contrasting approach here is to define separate data structures for each context, where the assumptions are encoded in the type itself, and translate between them at context boundaries, but this precludes code reuse (e.g. trait impls, methods), so there is a tradeoff here between code hygiene and code reuse that needs to be evaluated.
2. As shorthand for `assert!` (e.g. the equivalent of `assert!(binding.name.is_some())` in the code snippet you commented on), where the purpose of the assertion is sanity checking the internal logic of the program itself, and is not meant for end users to understand and deal with, i.e. the only fix is a new version of the program without the bug. In most cases, the implicit assertion is obvious, e.g. unwrapping an optional is an assertion that there is always a value at that point; but in cases where it isn't obvious, or has a more specific interpretation, `expect` is preferred.
3. Do I know what must have gone wrong to cause the implicit assertion fail? If yes, then use `expect`, and as a message, describe what must have happened in order for the panic to occur (e.g. `.expect("expected X to catch that this value is missing")`). If the panic could have occurred for a number of reasons, or the reason is inexplicable/impossible, then use `unwrap`, as the use of `expect` in such cases is more likely to be misleading/confusing than it is to be of assistance, but err on the side of `expect` if a useful hint can be provided.

So in this specific case, I'm applying those rules like so: 

1. We're reusing `TraceBinding` in two contexts: one in which the binding is concrete, and so by definition must have a name; and one in which the binding is abstract, and has no name. If we are operating in the concrete context, the name is guaranteed to be set.
2. We want to assert the guarantee described above, and simultaneously access the value we were promised is there; if that property doesn't hold, we have a pretty serious compiler bug
3. It should be impossible for `name` to be `None` in this context, and we have no idea how that could happen, so there is no useful descriptive text that would make `expect` preferable here. More important here is the backtrace and a copy of the input we blew up on

We don't have to get into that discussion here by the way, but figured I would elaborate on why I use `unwrap` as liberally as a I do in some of this code. It's not arbitrary, but it is based on an approach to some of the tradeoffs mentioned above that might be at odds with how those tradeoffs are evaluated by the rest of the team, e.g. maybe the team at large favors code hygiene over reuse. If that's the case, I can address the root cause rather than paper over the symptoms :)

Reply:
I think it would be good to get thoughts on this from @bobbinth

Reply:
My preference would be to use `.expect()` as much as possible - even if the message is trivial. Of course, it is better to have the message explain (briefly) why we expect a certain value to be available - but if that's not possible, I'd still prefer an `expect()` over an `unwrap()` in production code. Basically, the way we've been working so far is that if there is an `unwrap()`, the code is either for testing or benchmarking purposes.